lib/std/json.zig

```diff
@@ -1626,7 +1627,7 @@ fn parseInternal(
                     var r: T = undefined;
                     const source_slice = stringToken.slice(tokens.slice, tokens.i - 1);
                     switch (stringToken.escapes) {
-                        .None => mem.copy(u8, &r, source_slice),
+                        .None => if (r.len >= source_slice.len) mem.copy(u8, &r, source_slice) else return error.InsufficientSpace,
```

Comment:
I think this should also handle the case of the string being longer than the array. It also seems like the `unescapeValidString` below never validates the length of the string.

Reply:
Would it be more appropriate to do `output.len == source_slice.len`?

Is the string being longer than the array a non desired situation?

Reply:
Yes, just make sure the error name reflects that.

Reply:
Will do. For each use of `unescapeValidString` would something like this be acceptable?

`if (output.len == source_slice.len) try unescapeValidString(output, source_slice) else return error.LengthMismatch`

Reply:
and where `mem.copy` and `unescapeValidString` are used in the same switch statement like here 

https://github.com/ziglang/zig/blob/master/lib/std/json.zig#L1682..L1685

I'll add an if statement above it

```
if (output.len != source_slice.len) {
   return error.LengthMismatch;
}
switch (stringToken.escapes) {
   .None => mem.copy(u8, output, source_slice),
   .Some => try unescapeValidString(output, source_slice),
}
```

Reply:
`source_slice.len` is irrelevant if `stringToken.escapes != .None` other than that the check is fine.

Reply:
i think im getting confused. If `stringToken.escapes != .None` then `stringToken.escapes == .Some`. If `source_slice.len` is irrelevant when `stringToken.escapes == .Some` then what kind of string length validation should there be? If there should be string length validation what should it be compared to?

Reply:
Use `stringToken.decodedLength()` (which apparently is valid for strings without escapes too).

Reply:
Gotcha - thanks!

For here https://github.com/ziglang/zig/blob/master/lib/std/json.zig#L1628..L1631 I plan on doing

```
switch (stringToken.escapes) {
   .None => if (r.len == source_slice.len) mem.copy(u8, &r, source_slice) else return error.LengthMismatch,
   .Some => if (r.len == stringToken.decodedLength()) try unescapeValidString(&r, source_slice) else return error.LengthMismatch,
}
```

and for here https://github.com/ziglang/zig/blob/master/lib/std/json.zig#L1682..L1685 I plan on doing

```
switch (stringToken.escapes) {
   .None => if (output.len == source_slice.len) mem.copy(u8, output, source_slice) else return error.LengthMismatch,
   .Some => if (output.len == len) try unescapeValidString(output, source_slice) else return error.LengthMismatch,
}
```
where `len` is defined here https://github.com/ziglang/zig/blob/master/lib/std/json.zig#L1679

And for here https://github.com/ziglang/zig/blob/master/lib/std/json.zig#L1993 I plan on doing

`if (output.len == s.decodedLength()) try unescapeValidString(output, slice) else return error.LengthMismatch;`